adam/learner/propose_but_verify.py

```diff
@@ -0,0 +1,225 @@
+            # *match_ratio* is above the required value
+            if partial_match.match_ratio >= self._minimum_match_ratio:
+                logging.debug(
+                    f"Hypothesis for {concept} is confirmed with ratio: {partial_match.match_ratio}"
+                )
+                # We've verified our hypothesis so we don't need to learn anything from this scene.
+                # Note - We currently don't do any generalizing of our hypothesis space
+                # Which might be needed for anything other than objects
+                # Or the `minimum_match_ratio` value should be evaluated for those cases
+                # As the values from the pursuit paper are focused around objects
+                return
+
+        # If we either haven't seen this semantic template before or our hypothesis
+        # Wasn't confirmed We initialize a new concept or replace the current one.
+        if not concept:
+            concept = self._new_concept(
+                debug_string=bound_surface_template.surface_template.to_short_string()
+            )
+        self._surface_template_to_concept[
+            bound_surface_template.surface_template
+        ] = concept
+        self._concept_to_surface_template[
+            concept
+        ] = bound_surface_template.surface_template
+
+        self._concept_to_hypotheses[concept] = immutableset(
+            self._hypotheses_from_perception(
+                language_perception_semantic_alignment, bound_surface_template
+            )
+        )
+
+    def templates_for_concept(self, concept: Concept) -> AbstractSet[SurfaceTemplate]:
+        if concept in self._concept_to_surface_template:
+            return immutableset([self._concept_to_surface_template[concept]])
+        else:
+            return immutableset()
+
+    @abstractmethod
+    def _new_concept(self, debug_string: str) -> Concept:
+        """
+        Create a new `Concept` of the appropriate type with the given *debug_string*.
+        """
+
+    @abstractmethod
+    def _hypotheses_from_perception(
+        self,
+        learning_state: LanguagePerceptionSemanticAlignment,
+        bound_surface_template: SurfaceTemplateBoundToSemanticNodes,
+    ) -> AbstractSet[PerceptionGraphTemplate]:
+        """
+        Get a hypothesis for the meaning of *surface_template* from a given *learning_state*.
+        """
+
+    def _primary_templates(
+        self
+    ) -> Iterable[Tuple[Concept, PerceptionGraphTemplate, float]]:
+        return (
+            (concept, hypotheses[0], 1.0)
+            for (concept, hypotheses) in self._concept_to_hypotheses.items()
+            # We are confident in a hypothesis if we don't have any alternatives.
+            if len(hypotheses) == 1
+        )
+
+    def _fallback_templates(
+        self
+    ) -> Iterable[Tuple[Concept, PerceptionGraphTemplate, float]]:
+        # Alternate hypotheses stored in the beam.
+        return (
+            (concept, hypothesis, 1.0)
```

Comment:
Is there a reason we default to 1? I would think if these are fallback templates we'd want it to be lower but maybe that's not an issue?

Reply:
All confidence values are currently 1, it's available so algorithms can take advantage of it but we don't currently use it anywhere.